### PR TITLE
Ability to tag “bonus” levels in .script files for Stage Extras

### PR DESCRIPTION
**Pull request by joshlory**
Same approach as PR https://github.com/code-dot-org/code-dot-org/pull/8702.  Unfortunately these need to be an extra column for now, and require a migration.  In the future it may be worth moving `assessment`, `named_level` and `bonus` into the properties hash — I don't know of anywhere we query by those columns.

These `bonus` levels should be associated with specific stages, but should only appear in Stage Extras (not in the normal progression).  A future PR will add the Stage Extras bubble at the end of each stage.

### CODE REVIEW

**Comment by ashercodeorg** [comment]
Please make the migration a separate PR.

**Comment by Bjvanminnen** [comment]
"Unfortunately these need to be an extra column for now, and require a migration."

Why can't this just go in properties? Is it just because we want to be consistent with asessment/named_level?

File: dashboard/app/dsl/script_dsl.rb [hunk 1]
```diff
@@ -190,6 +195,7 @@ def self.serialize_stages(script)
         type = 'level'
         type = 'assessment' if sl.assessment
         type = 'named_level' if sl.named_level
+        type = 'bonus' if sl.bonus
```
Comment (Bjvanminnen) [inline dashboard/app/dsl/script_dsl.rb:198]: should we also have a test for the DSL changes?

**Comment by joshlory** [comment]
Yep, I'm following the existing pattern for consistency.

> In the future it may be worth moving `assessment`, `named_level` and `bonus` into the properties hash — I don't know of anywhere we query by those columns.

**Comment by Bjvanminnen** [comment]
I wonder a little bit whether this consistency has a higher value than starting to do things the right way. We may also be more likely to move assessment/named_level into properties once there's an example of the right way of doing things.

I don't think I feel very strongly about this, but my gut tells me we're better off putting this in properties.

**Comment by joshlory** [comment]
At risk of advocating for the lazy fix 😄, this will unblock Level Builders in the short term who want to start marking `bonus` levels that will be part of Stage Extras in the future.  I'd agree this is important tech debt to cleanup but right now I think the priority is higher to ship Stage Extras as a feature.

**Comment by ashercodeorg** [comment]
Though I don't (necessarily) disagree with this particular priority assessment, that we so often make this tradeoff is costing us in the long term. There needs to be more pushback on product on velocity requirements.

**Comment by joshlory** [comment]
+cc @ryansloan to assess velocity tradeoffs.  I don't feel super strongly here, but I think as long as we track the debt and pay it down after the Stage Extras feature ships I think we're ok.  That means actually doing the work, not pushing it out until "someday" (never) 😛.

I've added a Slack reminder to make sure I address this, and an item to the "Grace" sprint.

File: dashboard/db/migrate/20170427233825_add_bonus_to_script_level.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class AddBonusToScriptLevel < ActiveRecord::Migration[5.0]
+  def change
+    add_column :script_levels, :bonus, :boolean
```
Comment (Hamms) [inline dashboard/db/migrate/20170427233825_add_bonus_to_script_level.rb]: should this be non-nullable and default false?

Comment (joshlory) [reply dashboard/db/migrate/20170427233825_add_bonus_to_script_level.rb]: Ideally yes, but I'm following the pattern of the `assessment` & `named_level` columns.

**Comment by joshlory** [comment]
@ashercodeorg @Bjvanminnen any concerns with merging this as-is, with a plan to address the tech debt after the Stage Extras feature ships?

**Comment by Bjvanminnen** [comment]
:shipit:

**Comment by ashercodeorg** [comment]
Presumably you want to rebase to the migration PR, removing the migration changes from this PR?

Also, I would merge only after the migration has reached production.

**Comment by joshlory** [comment]
I verified https://github.com/code-dot-org/code-dot-org/pull/14758 has been deployed and I've merged staging into this branch.  Thanks!

**Pull request merged**